### PR TITLE
[HDRP] Add shadergraph nodes to sample custom pass color and depth buffers

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR

Added two new ShaderGraph nodes to sample the custom pass color and depth buffers in HDRP:

image

---
### Testing status

Manually tested all the depth sampling modes (raw, linear01, and linearEye) and custom color buffer on transparent objects (with a custom pass rendering in the before transparent injection point).

Added an automated test for both custom color and depth node:
image

https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/hd%252Fcustom-pass-buffer-shadergraph-nodes/.yamato%252Fall-hdrp.yml%2523PR_HDRP_2021.2/7800910/job

### CODE REVIEW

**Review by adrien-de-tocqueville — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Documentation~/Custom-Pass-Creating.md [hunk 1]
```diff
@@ -398,6 +398,8 @@ To change the buffer format of the Custom Pass component in your HDRP asset, go
 
 ![](Images/Custom_Pass_HDRPAsset_CBF.png)
 
+In Shader Graph, there are two nodes for sampling the custom color and depth buffers. For more information about these nodes, see the Shader Graph documentation: [Custom Color Node](https://docs.unity3d.com/Packages/com.unity.shadergraph@latest?subfolder=/manual/HD-Custom-Color-Node.html), [Custom Depth Node](https://docs.unity3d.com/Packages/com.unity.shadergraph@latest?subfolder=/manual/HD-Custom-Depth-Node.html).
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/Custom-Pass-Creating.md:401]: Thanks for adding this! Small suggestion: 

You can sample the custom color and depth buffers in Shader Graph using the following nodes: 
- [Custom Color Node](https://docs.unity3d.com/Packages/com.unity.shadergraph@latest?subfolder=/manual/HD-Custom-Color-Node.html)
- [Custom Depth Node](https://docs.unity3d.com/Packages/com.unity.shadergraph@latest?subfolder=/manual/HD-Custom-Depth-Node.html)

File: com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+| Name        | Direction           | Type  | Binding | Description |
+|:------------ |:-------------|:-----|:---|:---|
+| **UV** | Input | Vector 4 | Screen Position | Sets the normalized screen coordinates to sample. |
+| **Output** | Output      |    Vector 4 | None | Output value |
+
+## Depth Sampling modes
+| Name     | Description                        |
+|----------|------------------------------------|
+| Linear01 | Linear depth value between 0 and 1 |
```
Comment (Vic-Cooper) [inline com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md:21]: | Linear01 | The linear depth value between 0 and 1. |

File: com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+| **Node**       | **Universal Render Pipeline (URP)** | **High Definition Render Pipeline (HDRP)** |
+| -------------- | ----------------------------------- | ------------------------------------------ |
+| Custom Depth Node | No                                  | Yes                                        |
+
+## Ports
+
+| Name        | Direction           | Type  | Binding | Description |
+|:------------ |:-------------|:-----|:---|:---|
+| **UV** | Input | Vector 4 | Screen Position | Sets the normalized screen coordinates to sample. |
+| **Output** | Output      |    Vector 4 | None | Output value |
```
Comment (Vic-Cooper) [inline com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md:16]: | **Output** | Output      |    Vector 4 | None | The output value of this node. |

File: com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+| **Node**       | **Universal Render Pipeline (URP)** | **High Definition Render Pipeline (HDRP)** |
+| -------------- | ----------------------------------- | ------------------------------------------ |
+| Custom Depth Node | No                                  | Yes                                        |
+
+## Ports
+
+| Name        | Direction           | Type  | Binding | Description |
+|:------------ |:-------------|:-----|:---|:---|
+| **UV** | Input | Vector 4 | Screen Position | Sets the normalized screen coordinates to sample. |
```
Comment (Vic-Cooper) [inline com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md:15]: | **UV** | Input | Vector 4 | Screen Position | Sets the normalized screen coordinates that this node samples. |

File: com.unity.shadergraph/Documentation~/HD-Custom-Color-Node.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+| **Node**       | **Universal Render Pipeline (URP)** | **High Definition Render Pipeline (HDRP)** |
+| -------------- | ----------------------------------- | ------------------------------------------ |
+| Custom Color Node | No                                  | Yes                                        |
+
+## Ports
+
+| Name        | Direction           | Type  | Binding | Description |
+|:------------ |:-------------|:-----|:---|:---|
+| **UV** | Input | Vector 4 | Screen Position | Sets the normalized screen coordinates to sample. |
+| **Output** | Output      |    Vector 4 | None | The value in the custom pass color buffer at the sampled coordinates |
```
Comment (Vic-Cooper) [inline com.unity.shadergraph/Documentation~/HD-Custom-Color-Node.md:16]: | **Output** | Output      |    Vector 4 | None | The value the custom pass color buffer contains at the sampled coordinates. |

File: com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+| **Output** | Output      |    Vector 4 | None | Output value |
+
+## Depth Sampling modes
+| Name     | Description                        |
+|----------|------------------------------------|
+| Linear01 | Linear depth value between 0 and 1 |
+| Raw      | Raw depth value                    |
```
Comment (Vic-Cooper) [inline com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md:22]: | Raw      | The raw depth value.                    |

File: com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+| Eye      | Depth converted to eye space units |
```
Comment (Vic-Cooper) [inline com.unity.shadergraph/Documentation~/HD-Custom-Depth-Node.md:23]: | Eye      | The depth value converted to eye space units. |

**Review by Vic-Cooper — CHANGES_REQUESTED** [review]
A few small suggestions, generally good :)

**Review by sebastienlagarde — APPROVED** [review]

**Review by Vic-Cooper — APPROVED** [review]
Looks good :)

**Review by RemyUnity — APPROVED** [review]
Except the small issue I found and already corrected with commit f81656e everything seems fine.
I did set up a graph that allows to compare HDRP scene color & depth nodes VS the custom buffer nodes, and couldn't see any difference (when the custom pass is re-rendering the same things as the regular camera).
image

**Comment by sebastienlagarde** [comment]
For Shader GRaph team. The shield is due to documentation which is in shadergraph manual. Not a blocker. Merging.

**Pull request merged**